app/assets/javascripts/discourse/app/lib/highlight-syntax.js

```diff
@@ -1,34 +1,51 @@
+import loadScript from "discourse/lib/load-script";
+import deprecated from "discourse-common/lib/deprecated";
+
 /*global hljs:true */
 let _moreLanguages = [];
 
-import loadScript from "discourse/lib/load-script";
-
-export default function highlightSyntax($elem, siteSettings, session) {
+export default function highlightSyntax(elem, siteSettings, session) {
   const selector = siteSettings.autohighlight_all_code
-      ? "pre code"
-      : "pre code[class]",
-    path = session.highlightJsPath;
+    ? "pre code"
+    : "pre code[class]";
+  const path = session.highlightJsPath;
+
+  if (!elem) {
+    return;
+  }
+
+  if (elem instanceof jQuery) {
+    deprecated(
+      "highlightSyntax now takes a DOM node instead of a jQuery object.",
+      {
+        since: "2.6.0",
+        dropFrom: "2.7.0"
+      }
+    );
+
+    elem = elem[0];
+  }
 
   if (!path) {
     return;
   }
 
-  $(selector, $elem).each(function(i, e) {
-    // Large code blocks can cause crashes or slowdowns
-    if (e.innerHTML.length > 30000) {
-      return;
-    }
+  return loadScript(path).then(() => {
+    elem.querySelectorAll(selector).forEach(e => {
+      // Large code blocks can cause crashes or slowdowns
+      if (e.innerHTML.length > 30000) {
+        return;
+      }
 
-    $(e).removeClass("lang-auto");
-    loadScript(path).then(() => {
+      e.classList.remove("lang-auto");
       customHighlightJSLanguages();
       hljs.highlightBlock(e);
```

Comment:
I don't know how it affects the performance, but before the change it was running all `highlightBlock` concurrently (essentially `forEach(a => new Promise(() => highlightBlock(a)))`) and now it runs them synchronously though still within a promise (i.e. `new Promise(() => forEach(a => highlightBlock(a)))`)

Reply:
Hmmm yes indeed, at best it was an unknown desirable side-effect I would say. I doubt anyone would use the promise created by loadScript to make this more performant on purpose.